lix_2_92: init at 2.92.0

RaitoBezarius opened the pull request:
This brings the required stuff to package Lix 2.92 in Nixpkgs.

https://lix.systems/blog/2025-01-18-lix-2.92-release/

If you want to review this PR, to make it easier on authors, please get involved in Lix release management and do not mindlessly nixpkgs-review or comment on things you do not necessarily understand.

## Lix release management QA list

- [x] Build on 4 platforms
- [x] Build with debuginfo
- [x] Verify that manual is present
- [x] Verify that lix-doc is working as intended
- [x] Cross from x86_64 → aarch64
- [x] Cross from x86_64 → riscv64
- [x] Cross from x86_64 → armv7l
- [x] Cross from aarch64 → x86_64 (not mandatory)
- [ ] Static builds on x86_64-linux, aarch64-linux, aarch64-darwin: :red_circle: — broken!
- [x] Perform closure checks and verify that no unnecessary dependencies are included: Lix 2.92 got to 17MB (+2MB increase, mostly due to the Nix binary itself growing of 400 KB).
- [x] Ensure that previous versions did not explode in size neither in functionality:  all good (15MB).
- [x] Run various NixOS VM smoke tests: nixosTests.misc, etc. with Lix:
   - [x] Misc test (aarch64 & x86_64)
   - [x] Installer test

RaitoBezarius committed:
editline: apply patch to recognize alt-left & alt-right

This is required in the upcoming release of Lix and will benefit the
whole ecosystem of editline dependents.

It's already merged upstream and pending a final release.

Change-Id: I85fcf283494891c28bb05973bacb83030d942a9a
Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
editline: enable ANSI arrow keys and termcaps

This gives a richer editline for dependents in exchange of a ncurses
dependency.

This effectively makes editline not a replacement _without_ ncurses.
This can be re-obtained at the cost of a conditional or a feature flag
matrix like the kernel has.

Change-Id: Ic28c815171745e0e1fce2ffb6399f70490b7c359
Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
capnproto: support LLVM, tidy compilation flags, support debuginfo

The Lix team will take maintainership on this one as we are critically
dependent on the Cap'n'Proto runtime.

Change-Id: Iee99f88fa73540ea91a2ba2825fba1e0b8969a07
Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius committed:
lix_2_92: init at 2.92.0

This introduces https://lix.systems/blog/2025-01-18-lix-2.92-release/ in
nixpkgs.

This new release introduces a bunch of changes at the packaging level,
reliant on new features of Meson, new features of some of our
dependencies which are done in previous commits (editline & capnproto).

The approach taken here is that it's OK to add more dependencies than
required for older versions but not OK to break them.

Moreover, Lix is now load-bearing dependent on the Clang stdenv, it
cannot be compiled with GCC anymore.

An escape hatch is included but if this escape hatch is used, the user
is on its own and should probably work with GCC upstream to fix
miscompilations.

Change-Id: I4d51831c10e5d56723240b80fcafc9d951aff279
Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

RaitoBezarius commented on pkgs/tools/package-management/lix/common.nix:
```diff
@@ -4,14 +4,16 @@
     repo = "lix";
     rev = version;
-    inherit hash;
+    inherit hash postFetch;
```
will remove `postFetch`

RaitoBezarius commented on pkgs/tools/package-management/lix/default.nix:
```diff
@@ -69,6 +73,20 @@ lib.makeExtensible (self: {
+    common {
+      version = "2.92.0";
+      hash = "sha256-vm5Ddu2PFeu/zACE+M/xyT04sfZ4FApvyiUgrZ0BA84=";
+      docCargoHash = "sha256-waA1VX/xW9Yj+EBcu6bnAXgJN2pNkPfnIuV+cwxozc8=";
```
should get rid of this

RaitoBezarius commented on pkgs/tools/package-management/lix/default.nix:
```diff
@@ -69,6 +73,20 @@ lib.makeExtensible (self: {
+      hash = "sha256-vm5Ddu2PFeu/zACE+M/xyT04sfZ4FApvyiUgrZ0BA84=";
+      docCargoHash = "sha256-waA1VX/xW9Yj+EBcu6bnAXgJN2pNkPfnIuV+cwxozc8=";
+      cargoHash = "sha256-YMyNOXdlx0I30SkcmdW/6DU0BYc3ZOa2FMJSKMkr7I8=";
+      postFetch = ''
```
not required anymore

RaitoBezarius commented on pkgs/tools/package-management/lix/doc/default.nix:
```diff
@@ -1,14 +1,18 @@
   cargoLock ? null,
 }:
-
+let
```
all change of this file should disappear and this file should only be kept for <2.92 compat

getchoo commented on pkgs/tools/package-management/lix/common.nix:
With Clang being the default/requirement now, it'd probably be a good idea to implement [this](https://github.com/NixOS/nixpkgs/pull/353576/files#r1827189312) suggestion from https://github.com/NixOS/nixpkgs/pull/353576

I think something like the following should work and not effect previous versions. Everything built and tests ran just fine on my end
```diff
-      (lib.mesonBool "b_lto" (!stdenv.hostPlatform.isStatic && stdenv.cc.isGNU))
+      (lib.mesonBool "b_lto" (!stdenv.hostPlatform.isStatic && (lib.versionAtLeast version "2.92.0" && !stdenv.hostPlatform.isDarwin || stdenv.cc.isGNU)))
```

`pkgsLLVM.lix` should also be fixed after https://gerrit.lix.systems/c/lix/+/2286 (though I still wasn't able to build it since `lix-doc` is passing `-lgcc_s` to the compiler now for some reason, but applying the patch to previous versions worked)

RaitoBezarius replied:
Awesome work @getchoo!

RaitoBezarius replied:
Wait, @getchoo — I was definitely able to build with LTO on Darwin, what is happening?

RaitoBezarius replied:
Ah no, I'm tired, let me check again.

getchoo replied:
> I was definitely able to build with LTO on Darwin, what is happening?

LTO is currently disabled broadly on Clang due to https://github.com/NixOS/nixpkgs/issues/337036, which was theorized in https://github.com/NixOS/nixpkgs/pull/353576 to be the result of potential UB in Darwin or LLVM generally -- and since I've built Lix 2.91 against Clang + libcxx on Linux before, we know it can only be the former

I'm not sure if this still an in issue in 2.92 though, as I haven't built it. In the case it does build (against x86_64-darwin specifically) I'd imagine we're probably in the clear and this condition could be changed to a nicer `!stdenv.hostPlatform.isStatic && (lib.versionAtLeast version "2.92.0" || stdenv.cc.isGNU)`

lf- commented:
please hold for a lix 2.92.1, we found a critical bug that somehow we didn't manage to hit until after release

tahirmurata commented:
> please hold for a lix 2.92.1, we found a critical bug that somehow we didn't manage to hit until after release

Is there anywhere we can get more information?

lf- commented:
the first bug actually turned out to not affect 2.92.0 but we fixed it. currently blocked on https://git.lix.systems/lix-project/lix/issues/662

RaitoBezarius committed:
nixos/tests/installer: introduce `lix-simple` variant

An installer tests exacerbates the distribution packaging but in the
case of NixOS: the Nix package manager implementation.

As part of our classical release management process, the Lix team tests
whether a NixOS system installs just fine with Lix or not.

To avoid bloating the CI needlessly and keeping it simple, we only
introduce it on the simple variant and give a general way to pipe a Nix
implementation inside of an installer test.

Change-Id: I781da14475867dc2d946b740bad10af5de79ec5a
Signed-off-by: Raito Bezarius <masterancpp@gmail.com>

9999years commented on pkgs/by-name/ed/editline/package.nix:
```diff
@@ -31,12 +32,29 @@ stdenv.mkDerivation rec {
       hash = "sha256-m3jExTkPvE+ZBwHzf/A+ugzzfbLmeWYn726l7Po7f10=";
     })
+
+    # Pending editline new release:
+    #   https://github.com/troglobit/editline/pull/70
+    (fetchpatch {
+      name = "alt-left-and-alt-right.patch";
+      url = "https://github.com/troglobit/editline/commit/fb4d7268de024ed31ad2417f533cc0cbc2cd9b29.patch";
+      hash = "sha256-MDFe24lRDXKxEchACQXvC+Sw4sG0wpzBYdfe6UHa+iY=";
+    })
```
Done in #391426

9999years commented:
I think we're going to go with #393444 instead, which has been adapted from this PR now that parts of this have been merged in other PRs, and just not update `lixPackageSets.stable` to the new version on account of the `curl` bug.
